src/local/butler/reproduce.py

```diff
@@ -16,41 +16,114 @@
 from __future__ import print_function
 from builtins import object
 
+import httplib2
+import json
 import os
 import shutil
 import tempfile
+import urllib
+import webbrowser
 
+from src.python.base import utils
 from src.python.bot.tasks import commands
 from src.python.fuzzing import testcase_manager
 from src.python.system import environment
 from src.python.system import shell
 
+# TODO(mbarbella): This should work cross platform.
+AUTHORIZATION_CACHE_FILE = os.path.join(
+    os.path.expanduser('~'), '.config', 'clusterfuzz', 'authorization-cache')
 
-class _SimplifiedTestcase(object):
+# TODO(mbarbella): Don't use the old clusterfuzz-tools client id.
+OAUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth?%s' % (
+    urllib.urlencode({
+        'scope':
+            'email profile',
+        'client_id': ('981641712411-sj50drhontt4m3gjc3hordjmp'
+                      'c7bn50f.apps.googleusercontent.com'),
+        'response_type':
+            'code',
+        'redirect_uri':
+            'urn:ietf:wg:oauth:2.0:oob'
+    }))
+
+# TODO(mbarbella): This value should come from the configuration.
+TESTCASE_URL = 'https://clusterfuzz.com/reproduce-tool/testcase-info'
+
+
+class SerializedTestcase(object):
   """Minimal representation of a test case."""
 
-  def __init__(self, testcase_json):
-    self.crash_state = testcase_json['crash_state']
-    self.security_flag = testcase_json['security_flag']
-    self.gestures = testcase_json['gestures']
-    self.flaky_stack = testcase_json['flaky_stack']
+  def __init__(self, testcase_map):
+    self._testcase_map = testcase_map
 
-    # Custom field not included in real test cases. Used in environment setup.
-    self.job_definition = testcase_json['job_definition']
+  def __getattr__(self, item):
+    return self._testcase_map[item]
 
 
-def _get_testcase(_):
-  """Retrieve the json representation of the test case with the given id."""
-  # TODO(mbarbella): Actually fetch the test case info from the server.
-  testcase_json = {
-      'crash_state': '',
-      'security_flag': False,
-      'gestures': [],
-      'flaky_stack': False,
-      'job_definition': 'APP_NAME = echo\nAPP_ARGS = -n\n',
+class SuppressOutput(object):
+  """Suppress stdout and stderr.
+
+  We need this to suppress webbrowser's stdout and stderr."""
+
+  def __enter__(self):
+    self.stdout = os.dup(1)
+    self.stderr = os.dup(2)
+    os.close(1)
+    os.close(2)
+    os.open(os.devnull, os.O_RDWR)
+
+  def __exit__(self, *_):
+    os.dup2(self.stdout, 1)
+    os.dup2(self.stderr, 2)
+    return True
+
+
+def _get_authorization():
+  """Get the value for an oauth authorization header."""
+  # Try to read from cache.
+  authorization = utils.read_data_from_file(
+      AUTHORIZATION_CACHE_FILE, eval_data=False)
+  if authorization:
+    return authorization
+
+  # Prompt the user for a code if we don't have one.
+  with SuppressOutput():
```

Comment:
SupressOutput feels very hacky, can we just exclude it, it might be ok to have some stdout, stderr ?

Reply:
This is essentially just copied from the old tool, but not having it is problematic. It ends up outputting around the same time as the prompt for the verification code, which looks very confusing (essentially drowns out the request for the verification code).